css-view-transitions/navigation/hide-before-reveal.html is a flaky failure.

mattwoodrow opened the pull request:
#### dedbde02bdb86a53072012db36259448ba08dc1d
<pre>
css-view-transitions/navigation/hide-before-reveal.html is a flaky failure.
<a href="https://bugs.webkit.org/show_bug.cgi?id=278028">https://bugs.webkit.org/show_bug.cgi?id=278028</a>
&lt;<a href="https://rdar.apple.com/133772823">rdar://133772823</a>&gt;

Reviewed by Tim Nguyen.

This fails because the visiblity change steps from css-view-transitions-1
haven&apos;t been updated to account for cross-document view transitions yet.

It also clears inbound view transition params for cross-document view
transitions, which matches the requirements of the test and the behaviour of
blink - <a href="https://github.com/web-platform-tests/wpt/issues/47828.">https://github.com/web-platform-tests/wpt/issues/47828.</a> This should be
codified in the spec soon.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::visibilityStateChanged):

Canonical link: <a href="https://commits.webkit.org/283172@main">https://commits.webkit.org/283172@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2864ae7292d259572ea39606f13f35b10874bba1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65466 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44837 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18084 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69492 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/16075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67584 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/52638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/16355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/16075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68533 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/52638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/56662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/33196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/52638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/14037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14951 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/52638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/14378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/71197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/9420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/16355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/71197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/9452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/56723 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/71197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/1435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9924 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/40647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/41723 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/41467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2357,6 +2357,14 @@ void Document::visibilityStateChanged()
+            viewTransition->skipViewTransition(Exception { ExceptionCode::InvalidStateError, "Old view transition aborted by document becoming hidden."_s });
+    } else
+        ASSERT(!activeViewTransition());
 }
```
I made a patch here that does something similar: https://github.com/WebKit/WebKit/pull/26191

tuankiet65 replied:
Seems like @nt1m's previous patch is more complete? in the sense that it also reacts to the event of the page being made hidden.

mattwoodrow replied:
Mine should be doing that too. The difference is that @nt1m is a better citizen than I am, and used a visibility change listener, rather than hardcoding it into Document.

My reasoning is that a visibility change listener fires on the active view transition at the point of the visibility change, whereas the spec requires a task to be queued and the active view transition to be retrieved when the task runs.

Unfortunately, when I implemented the queued task, it causes tests to fail, and the assert to fire.

I followed up on my existing spec issue here: https://github.com/web-platform-tests/wpt/issues/47828#issuecomment-2316520253

Assuming the queueing of the task is just invalid, then @nt1m's patch is clearly better, and we should merge it (with the addition of the level 2 specific change `m_inboundViewTransitionParams = nullptr;`)

nt1m reviewed (approved):
Can we consider moving this logic in one place in ViewTransition.cpp?

mattwoodrow committed:
css-view-transitions/navigation/hide-before-reveal.html is a flaky failure.
https://bugs.webkit.org/show_bug.cgi?id=278028
<rdar://133772823>

Reviewed by Tim Nguyen.

This fails because the visiblity change steps from css-view-transitions-1
haven't been updated to account for cross-document view transitions yet.

It also clears inbound view transition params for cross-document view
transitions, which matches the requirements of the test and the behaviour of
blink - https://github.com/web-platform-tests/wpt/issues/47828. This should be
codified in the spec soon.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::visibilityStateChanged):

Canonical link: https://commits.webkit.org/283172@main
